### ISSUE TITLE
How to clone remote repo with --filter option?

### ISSUE THREAD

**Issue by RoxKilly** [go-git/go-git#713]
My golang application needs to list all file and folder paths in a subfolder of a git repo (a search path).  I only care about the HEAD commit.  I'm using `Clone()` first, then walking the tree from the HEAD commit.

```
repo, _ = git.Clone(memory.NewStorage(), nil, git.CloneOptions{URL: repoUrl, Depth: 1})
// Get commit object at HEAD
// get tree of commit object
// iterate into tree and append file names if file is within my search path
```

This works but the clone operation is very slow as the repo gets large and probably gives me a ton of stuff I don't need (I'm only interested in a subset of the repo tree).  I noticed that the [git clone docs][2] have a `--filter` option that could serve my purposes.  Is this option available via the go-git library? I don't see anything in the [CloneOptions object][3]

  [2]: https://git-scm.com/docs/git-clone
  [3]: https://github.com/go-git/go-git/blob/master/options.go#L37-L67

**Comment by AnthonyHerman** [comment]
I would also like to see this functionality come to go-git.

**Comment by jfontan** [comment]
You can set the `Depth` to 1 in `CloneOptions` to only download that commit.

**Comment by jfontan** [comment]
My bad. I had not noticed you are already using `Depth: 1`. Is the repository that takes so long public?

**Comment by pjbgf** [comment]
@RoxKilly `go-git` does not support [partial clones ](https://git-scm.com/docs/partial-clone) atm. But I would be happy to review a PR that implements it - so feel free to tag me if you decide to give it a go.

In the meantime, for a more efficient clone, you could change your `CloneOptions` to the following:

```
 &git.CloneOptions{
			URL:           repoUrl,
			SingleBranch:  true,
			Depth:         1,
			Tags: git.NoTags,
		})
```

More info as to why: https://github.com/go-git/go-git/issues/545#issuecomment-1353681676

**Comment by pjbgf** [comment]
Closing this as it seems that this feature is no longer supported upstream - at least not from a path perspective.

Git v2.22.0 [release notes](https://github.com/git/git/blob/master/Documentation/RelNotes/2.22.0.txt):
>  * The filter specification "--filter=sparse:path=<path>" used to  create a lazy/partial clone has been removed. 

The latest version of [git-rev-list](https://git-scm.com/docs/git-rev-list#_object_traversal) also states that:
> Note that the form `--filter=sparse:path=<path>` that wants to read from an arbitrary path on the filesystem has been dropped for security reasons.

Further details: https://gitlab.com/gitlab-org/git/-/issues/3

**Comment by santinoncs** [comment]
@pjbgf I need to implement 
```
git clone --filter=blob:none ...
```

so if is not a problem i would like to give it a go with a PR.

is that ok?

**Comment by santinoncs** [comment]
Where is the actual git clone happening in the code? can you guide me to that and i can continue from there?
thanks a lot
( i assume is where the fetch method )

**Comment by pjbgf** [comment]
@santinoncs pack protocol is mostly implemented [here](https://github.com/go-git/go-git/tree/master/plumbing/protocol/packp). For more information on how it works, I recommend reading the upstream documentation:

- https://github.com/git/git/blob/master/Documentation/gitprotocol-pack.txt
- https://github.com/git/git/blob/master/Documentation/gitprotocol-capabilities.txt

My understanding is that you need to implement the [filter](https://github.com/git/git/blob/bc5204569f7db44d22477485afd52ea410d83743/Documentation/gitprotocol-capabilities.txt#L373) capability. But I suspect this may depend on #264.

Reopening this for the implementation of filter options such as `--filter=blob:none`.

**Comment by cedricium** [comment]
For those curious, #1108 hopes to add the `--filter` to `FetchOptions` which would be pretty sweet if it lands.

**Comment by camball** [comment]
Is there anything holding #1108 up? I'd really love to use the equivalent functionality to `git clone <url> --filter=blob:none`, as it's the only way I can clone my large repo. My repo is very large and for the same reason as #982, I can't clone. After 6 or so minutes, the clone fails with `unexpected EOF`.

**Comment by pjbgf** [comment]
@camball please note that unless you have `Tags: git.NoTags`, all tags (and their references) will be cloned - here's an [example of the impact](https://github.com/go-git/go-git/issues/545#issuecomment-1353681676).

As for that given PR, I believe no maintainer had the time to go around testing / reviewing it yet. So I would recommend testing it yourself and see if it resolves your problem, then adding some comments to the PR to that effect, so that if they are issues the author can tackle them sooner rather than latter.

On the performance side of things, the `v6-exp` branch has improvements and better tracing. So running your application (or go-git's examples) with `GIT_TRACE_PERFORMANCE=true` should provide you further insights on what is consuming most of the time:

```
GIT_TRACE_PERFORMANCE=true go run _examples/clone/main.go https://github.com/go-git/go-git /tmp/git1
git clone https://github.com/go-git/go-git /tmp/git1 --recursive
08:54:52.128423 upload_pack.go:41: performance: 0.289098618 s: upload_pack
08:54:52.775860 common.go:32: performance: 0.646588807 s: update_obj_storage
08:54:53.094876 worktree.go:330: performance: 0.311227125 s: reset_worktree
08:54:53.094971 repository.go:489: performance: 1.497164190 s: git command: git clone https://github.com/go-git/go-git
...
```

If performance issues persist, please open a new issue with the steps to reproduce.

**Comment by camball** [comment]
@pjbgf That's excellent! `Tags: git.NoTags` worked beautifully, reducing clone time from the 6 minute mark down to around a minute. My eyes glazed right over the `Tags` option. Thanks :)

**Comment by pjbgf** [comment]
Closing this as per #1375. A follow-up [issue](#1381) was created for full support of partial clones.

**Comment by traut** [comment]
as the link in the last comment is not correct here is a follow-up issue -- https://github.com/go-git/go-git/issues/1381
